Split identity allocation into pkg/allocator and abstracted Backends

raybejjani opened the pull request:
We intend to add CRD backed Identities (https://github.com/cilium/cilium/pull/8236) and that requires some refactoring. This PR refactors the allocation business logic so that it is collected into pkg/allocator. The existing kvstore specific logic remains in pkg/kvstore/allocator. Note that pkg/kvstore supports multiple backends of its own, consul or etcd. `pkg/kvstore/allocator.Allocator` implements (and is) a pkg/allocator.Backend. This is used by the node-global identity allocator.

Overall, nothing should behave differently than before.

Some functions on allocator.Backend are renamed versions of originals:
`func AllocateID(ctx context.Context, id idpool.ID, key AllocatorKey) error` from 
`func (a *Allocator) recreateMasterKey(id idpool.ID, value string, reliablyMissing bool)`
Note that Allocate itself, the function, mostly lives in pkg/allocator.Allocator and is roughly the same. It called `recreateMasterKey`via lockedAllocate.

`func AcquireReference(ctx context.Context, id idpool.ID, key AllocatorKey, lock kvstore.KVLocker
) error` from
`func (a *Allocator) createValueNodeKey(ctx context.Context, key string, newID idpool.ID, lock kvstore.KVLocker) error`
In fact, it just wraps it. There is logic in `createValueNodeKey` that dealt with the `cache.verify` to record successful ID allocation and it is now in `pkg/allocator.Allocator::lockedAllocate`. This is one of 2 call-sites for AcquireReference (both in lockedAllocate) and they were the same for `createValueNodeKey`

`ListAndWatch(handler CacheMutations, stopChan chan struct{})` and `CacheMutations` abstract out callbacks from `localkeys` that are made on create/modify/delete of keys. This code was in cache.go::start but is delegated to the `Backend` now (and takes CacheMutations instead of an Allocator because it's import loops the whole way down).

2019-07-13
Rebased onto master
- Merged all temporary commits in case we merge it. I have the older commits if anyone wants a copy to review.
- Added a comment to the commented tests in pkg/allocator/allocator_test.go (moved from pkg/kvstore/allocator/allocator_test.go) explaining that they no longer compile and how to fix that.

2019-07-12
Rebased onto master
- removed a double copy of some commented out tests for allocator (from https://github.com/cilium/cilium/blob/master/pkg/kvstore/allocator/allocator_test.go#L448-L528). They remain commented out and will no longer compile (they test multi-node allocation and the dummyBackend assumes it is a single node.)
- Added a missing localKeys.verify when allocating a node (it is a separate commit). This logic was in createValueNodeKey previously, and wasn't correctly brought over to all call-sites of the replacement function (AcquireReference).
- Split comments to reflect Allocator/Backend split.

2019-07-11
Rebased onto master.
~~I've added two commits that add locking onto all the allocator.Backend functions. These are intended to be squashed, but might be easier to review if anyone has read the earlier commits.~~
I've added two commits that add locking onto some allocator.Backend functions that are needed for pkg/allocatore.Allocator.lockedAllocate to do what it did before(the second of the commits). This should be the same behaviour as before, although UpdateKey technically does more when the `reliablyMissing` parameter is true. It is false in the `lockedAllocate` case. `lockedAllocate` came from `pkg/kvstore/allocator` if you want to compare.
- renamed TestType -> TestAllocatorKey in tests in pkg/allocator and pkg/kvstore/allocator. This was merged into the main commit
- corrected pkg/allocator.Allocator::syncLocalKeys to only log on error not return early (the old behaviour was to ignore errors with no logging). This was merged into the main commit.
- added some commentary (above) about functions that have been renamed but have older equivalents.

2019-07-10
Rebased onto master.
I've split the main commit into a "copy the code to the new package" commit and the original one. Hopefully this makes it a little easier to review. The tests aren't too amendable to this so in the copy code commit they are just commented out to avoid import loops.

2019-07-09
Rebased onto master.
Small fixups that have been merged in.


<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8496)
<!-- Reviewable:end -->

tgraf committed:
kvstore: Abstract identity allocator backends

We intend to add CRD backed Identities and that requires some
refactoring. This change refactors the allocation business logic so that it
is collected into pkg/allocator. The existing kvstore specific logic
remains in pkg/kvstore/allocator. Note that pkg/kvstore supports
multiple backends of its own, consul or etcd.
pkg/kvstore/allocator.Allocator implements (and is) a
pkg/allocator.Backend. This is used by the node-global identity
allocator.

Overall, nothing should behave differently than before.

Co-authored-by: Ray Bejjani <ray@covalent.io>
Signed-off-by: Ray Bejjani <ray@covalent.io>

raybejjani committed:
allocator: keyToID no longer deletes invalid keys

This was a little unexpected. This change factors out the invalid key
handling to the only location that needed it. There should be no
functionality change.

Signed-off-by: Ray Bejjani <ray@covalent.io>

ianvernon commented on pkg/kvstore/allocator/cache.go:
```diff
@@ -160,8 +158,16 @@ func (c *cache) start(a *Allocator) waitChan {
+						kvstore.Delete(event.Key)
+					}
+
+				case id != 0:
```
0 --> NoID

ianvernon commented on pkg/kvstore/allocator/cache.go:
```diff
@@ -160,8 +158,16 @@ func (c *cache) start(a *Allocator) waitChan {
+				id, err := k.keyToID(event.Key)
+				switch {
+				case err != nil:
+					log.WithError(err).Warning("Invalid key")
```
Add a field indicating what the key is.

ianvernon reviewed (commented):
Reviewing the first commit, another review incoming for the second commit...

ianvernon commented on operator/identity_gc.go:
```diff
@@ -17,19 +17,28 @@ package main
 import (
```
Not entirely sure what the correct / best practice here is, but should the commit include @tgraf as a co-author / have his sign-off(?)

raybejjani replied:
tgraf was still the author on the commit but I added myself as co-author (at least, I think I did in a way github will like). I can switch it the other way around if anyone prefers.

ianvernon commented on pkg/allocator/allocator.go:
```diff
@@ -0,0 +1,787 @@
+// 2. Do Get() on master key, return result
+//
+// Allocate:
+// 1. Check local key cache, increment, and return if key is already in use
```
Isn't this specific to the kvstore implementations, as opposed to the CRD-backed allocator? This won't be the case for the CRD-backed allocator based on what you mentioned about how the we leverage the labels fields within the CRD itself.

raybejjani replied:
This section isn't, but the nomenclature is a bit different. In particular, slave/master key may not quite apply anymore (we use reference in the Backend interface, for example). The section above, however, is kvstore specific. I'll split this up and rewrite it a bit.

ianvernon commented on pkg/allocator/allocator.go:
```diff
@@ -0,0 +1,787 @@
+//
+// 1.1. If found, increment and return (no kvstore interactions)
+// 2. Lookup ID by key in local cache or via first slave key found in kvstore
+// 2.1
```
Can 2.1 and 3 be removed?

raybejjani replied:
perhaps they were reserved for future use? ;P
I'll remove them.

ianvernon commented on pkg/allocator/allocator.go:
```diff
@@ -0,0 +1,787 @@
+// Release:
+//  1. Reduce local reference count until last use (no kvstore interactions)
+//  2. Delete slave key (basePath/value/key1/node1)
+//     This automatically guarantees that when the last node has relesed the
```
Nit: relesed --> released

ianvernon commented on pkg/allocator/allocator.go:
```diff
@@ -0,0 +1,787 @@
+// AllocatorOption is the base type for allocator options
+type AllocatorOption func(*Allocator)
+
+// NewAllocatorForGC returns an allocator  that can be used to run RunGC()
```
Nit: Remove the extra space between "allocator" and "that"

tgraf commented on pkg/kvstore/allocator/cache.go:
```diff
@@ -160,8 +158,16 @@ func (c *cache) start(a *Allocator) waitChan {
+				case err != nil:
+					log.WithError(err).Warning("Invalid key")
+
+					if k.deleteInvalidPrefixes {
```
`invalidKey()` is now unused and can be removed.

nebril commented on operator/identity_gc.go:
```diff
@@ -17,19 +17,28 @@ package main
+
+	// identityAllocationMode specifies what mode to use for identity
+	// allocation
+	identityAllocationMode string
```
This does not seem to be used anywhere.

ianvernon commented on pkg/kvstore/client.go:
```diff
@@ -30,7 +30,7 @@ func initClient(module backendModule, opts *ExtraOptions) error {
 	if c == nil {
 		err := <-errChan
-		log.WithError(err).Fatalf("Unable to create etcd client")
+		log.WithError(err).Fatalf("Unable to create kvstore %s client", module.getName())
```
Nit: use `WithFields` here for a generic log message with the field being the output of `module.getName()`.

ianvernon commented on pkg/labels/labels.go:
```diff
@@ -384,6 +384,16 @@ func NewLabelsFromSortedList(list string) Labels {
 }
 
+// NewLabelArrayFromMap creates labels from a map
+func NewLabelArrayFromMap(base map[string]string, source string) LabelArray {
```
This appears to be new functionality - why is this needed as part of this refactor? Factor this out into a separate commit and add additional unit tests for it.

raybejjani replied:
yup! It's actually only used in the follow-up PR (https://github.com/cilium/cilium/pull/8236/) but, in looking at this, I realised that it isn't directly needed. There are existing functions that can be composed to do it, albeit less efficiently. I'll go with that instead, since less code is better :)

aanm commented on pkg/identity/cache/allocator.go:
```diff
@@ -47,6 +48,17 @@ func (gi globalIdentity) GetKey() string {
+func (gi globalIdentity) GetAsMap() map[string]string {
+	m := map[string]string{}
+
+	for _, v := range gi.LabelArray {
```
Consider making this function as a method of `LabelArray`

raybejjani replied:
good idea. I noticed that Labels has a function called StringMap that is exactly this code. I added a StringMap to LabelArray. Not exactly what you had in mind but I didn't want to have two different function names for the same thing. What do you think? (I'll push it up with my other changes soon)

aanm commented on pkg/kvstore/allocator/allocator.go:
```diff
@@ -16,27 +16,23 @@ package allocator
 
 var (
-	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "allocator")
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "kvstorebackend")
```
It's just a little odd the filename not corresponding to the subsys field

aanm commented on pkg/kvstore/allocator/allocator.go:
```diff
@@ -821,128 +374,82 @@ func (a *Allocator) RunGC(staleKeysPrevRound map[string]uint64) (map[string]uint
-	case recreated:
-		log.WithField(fieldKey, keyPath).Warning("Re-created missing master key")
+func (k *kvstoreBackend) keyToID(key string) (id idpool.ID, err error) {
+	if !strings.HasPrefix(key, k.idPrefix) {
```
Why is this important? Why does the key require a prefix? Won't be simpler if the key didn't contain a prefix?

raybejjani replied:
It was like this before. The places we call this use the key from the kvstore directly, so it would contain the full path. The prefix here would be the configured prefix for this whole instance (so `/cilium/state` or something similar) since we watch for events on master and slave keys (`/cilium/state/id/...` and `cilium/state/value/...`). This is the function that converts to the numeric ID.
```
func (c *cache) keyToID(key string, deleteInvalid bool) idpool.ID {
        if !strings.HasPrefix(key, c.prefix) {
               invalidKey(key, c.prefix, deleteInvalid)
               return idpool.NoID
        }

        suffix := strings.TrimPrefix(key, c.prefix)
        if suffix[0] == '/' {
                suffix = suffix[1:]
        }

       id, err := strconv.ParseUint(suffix, 10, 64)
        if err != nil {
               invalidKey(key, c.prefix, deleteInvalid)
               return idpool.NoID
        }

       return idpool.ID(id)
 }
```

aanm reviewed (commented):
nothing blocking

ianvernon reviewed (commented):
More comments - still need to plumb through review.

ianvernon commented on pkg/allocator/localkeys.go:
```diff
@@ -51,24 +51,26 @@ func newLocalKeys() *localKeys {
-func (lk *localKeys) allocate(key string, val idpool.ID) (idpool.ID, error) {
+// Note: keyString is expected to be key.GetKey(). We allow it to be passed in
+// to avoid creating large strings repeatdly.
+func (lk *localKeys) allocate(keyString string, key AllocatorKey, val idpool.ID) (idpool.ID, error) {
```
This looks like a behavioral change vs. a refactor, is that correct? Please factor it out into a separate commit if possible. This patch is huge and having to look at what is functionally changed vs. moved into separate packages is hard :/

raybejjani replied:
This specific function is almost exactly the same. The only difference is that the key was converted from a string to AllocatorKey. We also pass in `keyString` (I assume to avoid recreating the string repeatedly, as callers of this function use the key as a string a few times).
Did you mean that key is no longer a string? I could try to undo the abstraction but I'm not sure how much of it is removable this way. In any case, I've removed the extra parameter, so the only change is that now `key` is an `AllocatorKey`.

ianvernon replied:
My main concern is that there are small changes interspersed throughout code-shuffling, which means I may gloss over something which is small, but significant, in terms of behavior. Being able to know for sure that the only thing that has changed in a patch is the location of code rather than the code itself, is useful so that functionality changes can be analyzed more closely.

raybejjani replied:
Ok, @ianvernon I did manage an intermediate commit (https://github.com/cilium/cilium/pull/8496/commits/736725f69dd7eecdaf6575461249510008f8ee69) that should be only the code copies. It isn't quite correct, though. The tests are commented out completely because they caused import loops that I couldn't fix. Most of the differences should be renaming things to be in new packages, but it isn't all so straighforward. Does it help at all?

ianvernon commented on pkg/allocator/localkeys.go:
```diff
@@ -99,15 +101,15 @@ func (lk *localKeys) lookupKey(key string) idpool.ID {
 
 // lookupID returns the key for a given ID or an empty string
-func (lk *localKeys) lookupID(id idpool.ID) string {
+func (lk *localKeys) lookupID(id idpool.ID) AllocatorKey {
```
Same here.

ianvernon commented on pkg/identity/cache/allocator.go:
```diff
@@ -47,6 +48,17 @@ func (gi globalIdentity) GetKey() string {
 }
 
+// GetAsMap() encodes a globalIdentity as string
+func (gi globalIdentity) GetAsMap() map[string]string {
```
Add unit tests here to ensure correctness.

raybejjani replied:
I did this indirectly by testing LabelArray.StringMap, itself a copy of Labels.StringMap which is the same code as this but on the other label collections class.

ianvernon commented on pkg/kvstore/allocator/allocator.go:
```diff
@@ -643,9 +180,9 @@ func (a *Allocator) GetNoCacheIfLocked(ctx context.Context, key AllocatorKey, lo
-	pairs, err := kvstore.ListPrefixIfLocked(prefix, lock)
-	kvstore.Trace("ListPrefixLocked", err, logrus.Fields{fieldPrefix: prefix, "entries": len(pairs)})
+	prefix := path.Join(k.valuePrefix, key.GetKey())
+	pairs, err := kvstore.ListPrefix(prefix)
```
Why the change from `ListPrefixIfLocked` back to `ListPrefix`? This appears to be a functional change.

raybejjani replied:
I need to double check this. There was a reason I thought it ok that locking wasn't here but I shall revisit it.

raybejjani replied:
ok, @ianvernon I redid the code. All functions have locking where the old code did. These changes are in 2 commits, one for the Get* functions and another for the remainder. They're intended to be squashed but I figured it was easier to see the addition this way.
@aanm you might also want to look at this since you added some of the locking in lockedAllocate, I think.

ianvernon replied:
Thanks for doing that, will look at those commits!

ianvernon commented on pkg/kvstore/allocator/allocator.go:
```diff
@@ -679,9 +217,9 @@ func (a *Allocator) GetNoCache(ctx context.Context, key AllocatorKey) (idpool.ID
-	pairs, err := kvstore.ListPrefix(prefix)
-	kvstore.Trace("ListPrefix", err, logrus.Fields{fieldPrefix: prefix, "entries": len(pairs)})
+	prefix := path.Join(k.valuePrefix, key.GetKey())
+	pairs, err := kvstore.ListPrefixIfLocked(prefix, lock)
```
Here as well.

ianvernon commented on pkg/kvstore/allocator/allocator.go:
```diff
@@ -700,67 +238,82 @@ func (a *Allocator) GetNoCache(ctx context.Context, key AllocatorKey) (idpool.ID
-
-	v, err := kvstore.Get(path.Join(a.idPrefix, id.String()))
+func (k *kvstoreBackend) GetByID(id idpool.ID) (allocator.AllocatorKey, error) {
+	v, err := kvstore.Get(path.Join(k.idPrefix, id.String()))
```
Why do we have to query the kvstore directly vs. accessing the cache now? This seems like a functional change, unless the callees of `GetByID` now do the cache lookup before calling `GetByID`?

raybejjani replied:
In this case, this is the kvstore. The only caller, in reality, is pkg/allocator.Allocator. In that GetByID, it does check the cache first.

ianvernon commented on pkg/kvstore/allocator/allocator_test.go:
```diff
@@ -453,89 +420,6 @@ func (s *AllocatorSuite) TestRemoteCache(c *C) {
+	a.Delete()
 }
-
-// The following tests are currently disabled as they are not 100% reliable in
```
Why was this deleted?

raybejjani replied:
hmm, I thought these ended up in pkg/allocatore/allocator_test.go but apparently not, I'll put them back in.

ianvernon commented on pkg/allocator/allocator_test.go:
```diff
@@ -0,0 +1,334 @@
+// Copyright 2016-2017 Authors of Cilium
```
Change copyright to 2019

ianvernon commented on pkg/allocator/cache.go:
```diff
@@ -0,0 +1,229 @@
+	c.nextCache[val] = key
+	c.nextKeyCache[key.GetKey()] = val
+	c.mutex.Unlock()
+}
```
Are there any notable code changes in this file vs. just moving to the new package? Would be helpful to have them pointed out if so.

raybejjani replied:
Unfortunately, yes. I'll mark them above, but I'll also try to redo the PR with a commit to copy code between packages and another that creates the abstraction.

ianvernon commented on pkg/kvstore/allocator/allocator.go:
```diff
@@ -134,501 +63,109 @@ type Allocator struct {
-	mainCache cache
+	// deleteInvalidPrefixes enables deletion of identities outside of the
+	// valid prefix
+	deleteInvalidPrefixes bool
```
To other reviewers - this was moved here because the `mainCache` field is now in the allocator, not in the kvstore backend.

ianvernon reviewed (commented):
Some additional comments. It'd be helpful to note via comments on functions whether there are any changes beyond copy / paste / re-package given that the delta is quite big. The interfaces as defined + the splitting out appears OK, though, but I may be missing some subtle changes due to the PR's size.

raybejjani committed:
labels: Add LabelArray.StringMap function

Labels.StringMap converts a Labels into a map[string]string. There are
cases where similar behaviour is needed with a LabelArray.
LabelArray.StringMap is almost the exact same code as the Labels
variant, but it operates on a LabelArray.
Note that in the Labels case the Labels strucure implicitly deduplicates
on key, ignoring namespace. LabelArray is simply an array.

Signed-off-by: Ray Bejjani <ray@covalent.io>

nebril commented on pkg/kvstore/allocator/allocator_test.go:
```diff
@@ -70,72 +71,40 @@ func (e *AllocatorConsulSuite) TearDownTest(c *C) {
 	kvstore.Close()
 }
 
+//FIXME: this should be named better, it implements pkg/allocator.Backend
```
Will `TestAllocatorBackend` work for a name?

raybejjani replied:
oh, I completely forgot about this. I'll take that name :)

nebril commented on pkg/allocator/allocator.go:
```diff
@@ -0,0 +1,788 @@
+				log.WithError(err).Warning("Unable to run local key sync routine")
+			}
+
+			select {
```
Should this select deal with context timing out/ending? I guess in that case we should create the context in `NewAllocator` function.

Otherwise maybe remove `ctx` arg from `syncLocalKeys` and pass `context.TODO()` directly to `a.Backend.UpdateKey` call on line 712?

Basically I think this place is weird for suddenly starting using context, please disregard if it doesn't make sense to you.

raybejjani replied:
hmm, yeah. I was plumbing in the context and then I hit this function and it was difficult so I stopped. This was mostly because I wanted to be able to cancel the sync operation (assuming it's supported all the way down the stack, but that is a lie I think).
I just realised though, that I can do this:
```
ctx, cancel := context.WithCancel(context.TODO())
...
if err := a.syncLocalKeys(ctx); err != nil {
...
			case <-a.stopGC:
                                 cancel()
				log.Debug("Stopped master key sync routine")
				return
```

This way, stopGC is wired through to the lower levels. An alternative would be to defer the cancel. What do you think? too obtuse?

nebril replied:
What are we gaining here with the `cancel` call? Since AFAIU we don't have anything (resources/state) associated with this context, cancelling it is basically a no-op at this point, right?

raybejjani replied:
ok. I undid this completely.

nebril commented on pkg/kvstore/allocator/allocator.go:
```diff
@@ -700,67 +238,82 @@ func (a *Allocator) GetNoCache(ctx context.Context, key AllocatorKey) (idpool.ID
+// UpdateKey refreshes the record that this node is using this key -> id
+// mapping. When reliablyMissing is set it will also recreate missing master or
+// slave keys.
+func (k *kvstoreBackend) UpdateKey(ctx context.Context, id idpool.ID, key allocator.AllocatorKey, reliablyMissing bool) error {
```
Nitpick - `reliablyMissing` isn't clear for me. What about `forceUpdate`? I see that this name was used before so feel free to ignore this comment.

raybejjani replied:
This was the name before, when this function was called `recreateMasterKey` in pkg/kvstore/allocator so I won't change it unless all the other reviewers also want it. It is a confusing name nonetheless.

tgraf replied:
reliablyMissing is true when it is known that the key is missing and it's not an opportunistic update. This is the case when we re-create after having received a delete notification for a key that is owned by the local agent.

ianvernon commented on pkg/kvstore/allocator/allocator.go:
```diff
@@ -131,6 +131,18 @@ func (k *kvstoreBackend) AllocateID(ctx context.Context, id idpool.ID, key alloc
+	// create /id/<ID> and fail if it already exists
+	keyPath := path.Join(k.idPrefix, id.String())
+	success, err := kvstore.CreateOnlyIfLocked(ctx, keyPath, []byte(key.GetKey()), false, lock)
+	if err != nil || !success {
```
Maybe `switch` and have a different error returned depending on if the error is non-nil or `!success` since if the error is nil, but `success == false` we will return a nil string in the error we create.

raybejjani replied:
I did start down that road but the original code conflates `err != nil` and `!success` so I left it as-is.  Looking at the pkg/kvstore/etcd.go  implementation it seems like `!success` is always `err != nil` (there are hardcoded `return false, err` statements).
https://github.com/cilium/cilium/blob/master/pkg/kvstore/allocator/allocator.go#L460-L463
```
			success, err := kvstore.CreateOnlyIfLocked(ctx, keyPath, []byte(k), false, lock)
			if err != nil || !success {
				return 0, false, fmt.Errorf("unable to create master key '%s': %s", keyPath, err)
			}
```

ianvernon commented on pkg/kvstore/allocator/allocator_test.go:
```diff
@@ -16,526 +16,411 @@
 //
-//func testParallelAllocator(c *C, maxID idpool.ID, allocatorName string, suffix string) {
-//	allocator, err := NewAllocator(allocatorName, TestType(""), WithMax(maxID), WithSuffix(suffix))
+//type AllocatorEtcdSuite struct {
```
Add a comment indicating why these are commented out. It should be a priority for us to re-enable these tests somehow soon so we don't lose critical coverage for the kvstore.

raybejjani replied:
I copied the comment that was already there. I actually don't know what was exactly wrong, but they are ""not 100% reliable". I'll include a comment that they were copied over and may no longer even compile. I looked through them and they would need a rewrite of the dummyBackend because they test multi-node access. I'll try to sort this out on the plane.

raybejjani committed:
kvstore: Split logic into pkg/allocator

This is an intermediate commit. Some logic from pkg/kvstore/allocator has
been moved into pkg/allocator. No functionality should change, and most
changes are there to correct for package changes.
Unfortunately, the test logic is difficult to refactor this way. I have
commented it out since it is fixed in the next commit.

Signed-off-by: Ray Bejjani <ray@covalent.io>
